incubator/etcd/templates/statefulset.yaml

```diff
@@ -3,198 +3,262 @@ kind: StatefulSet
-        imagePullPolicy: "{{ .Values.ImagePullPolicy }}"
-        ports:
-        - containerPort: {{ .Values.PeerPort }}
-          name: peer
-        - containerPort: {{ .Values.ClientPort }}
-          name: client
-        resources:
-          requests:
-            cpu: "{{ .Values.Cpu }}"
-            memory: "{{ .Values.Memory }}"
-        env:
-        - name: INITIAL_CLUSTER_SIZE
-          value: {{ default 3 .Values.Replicas | quote }}
-        - name: SET_NAME
-          value: {{ template "etcd.fullname" . }}
-        volumeMounts:
-        - name: datadir
-          mountPath: /var/run/etcd
-        lifecycle:
-          preStop:
-            exec:
-              command:
-                - "/bin/sh"
-                - "-ec"
-                - |
+        - name: {{ .Chart.Name }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          ports:
+            - name: peer
+              containerPort: 2380
+            - name: client
+              containerPort: 2379
+          env:
+            - name: INITIAL_CLUSTER_SIZE
+              value: "1"
+            - name: SET_NAME
+              value: {{ template "etcd.fullname" . }}
+            - name: POD_NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+          volumeMounts:
+            - name: datadir
+              mountPath: /etcd
+          lifecycle:
+            preStop:
+              exec:
+                command:
+                  - "/bin/sh"
+                  - "-ecx"
+                  - |
+                    # Cleaning up after a controlled stop, like a scale down of the cluster
+                    # is needed so the member is removed and could be started fresh
+                    {{- if .Values.etcd.peerTLS }}
+                    PROTO=https
+                    {{- else }}
+                    PROTO=http
+                    {{- end }}
+                    EPS=""
+                    for i in $(seq 0 $((${INITIAL_CLUSTER_SIZE} - 1))); do
+                        EPS="${EPS}${EPS:+,}http://${SET_NAME}-${i}.${SET_NAME}.${POD_NAMESPACE}:2379"
+                    done
+                    HOSTNAME=$(hostname)
+                    ETCD_DIR="/etcd"
+                    member_hash() {
+                        etcdctl member list | grep ${PROTO}://${HOSTNAME}.${SET_NAME}.${POD_NAMESPACE}:2380 | cut -d':' -f1 | cut -d'[' -f1
+                    }
+                    echo "Removing ${HOSTNAME} from etcd cluster"
+                    ETCDCTL_ENDPOINT=${EPS} etcdctl member remove $(member_hash)
+                    if [ $? -eq 0 ]; then
+                        # Remove everything otherwise the cluster will no longer scale-up
+                        rm -rf $ETCD_DIR
+                    fi
+          command:
+            - "/bin/sh"
```

Comment:
Lots of initialisation here, I would recommend moving it to an initContainer instead, unless you have a good reason to keep it here? If you do an initContainer, I would also recommend to use a ConfigMap with this script to run, as it's easier to read.

Reply:
Would require the `initContainer` to pass some information about the state of `etcd` to the actual container as the initialization script does some state determinations before it launches the `etcd` daemon with the correct options. I do agree that it would improve readability.